text/trace/0192-messaging-semantic-conventions-spec.md

```diff
@@ -0,0 +1,165 @@
+
+   This layer helps to model dependencies between producers and consumers,
+   regardless of the underlying messaging transport mechanism and its
+   instrumentation.
+2. The _transport context layer_ allows correlating the producer and the
+   consumer with an intermediary. It also allows to correlate multiple
+   intermediaries among each other. The transport context can be changed by
+   intermediaries, according to intermediary instrumentations. Intermediaries
+   that are not instrumented might simply drop the transport context.
+
+   This layer helps to gain insights into details of the message transport.
+
+A producer SHOULD attach a creation context to each message. The creation context
+SHOULD be attached in a way so that it is not changed by intermediaries. A
+producer MAY propagate a transport context to an intermediary.  An
+intermediary MAY propagate a transport context to a consumer.
+
+### Trace structure, names, and attributes
+
+#### Consumer
+
+For many use cases, it is not possible to rely on the presence of "Process"
+spans for correlating producer with consumer traces: there are cases where a
+dedicated processing operation cannot be identified, or where processing
+happens in a different trace. Furthermore, processing operations often are not
+covered by messaging libraries and SDKs, but take place in application code.
+Consistently creating spans for "Processing" operations would require either
+effort from the application owner to correctly instrument those operations, or
+additional capabilities of messaging libraries and SDKs (e. g. hooks for
+processing callbacks, which can then be instrumented by the libraries or SDKs).
+
+While it is possible to create "Process" spans and correlate those with
+consumer traces in certain cases, this is not something that can be generally
+required. Therefore, it is more feasible to require the creation of "Deliver"
+spans to correlate producer with consumer traces. A "Deliver" span must link
+to the "Create" spans of all messages that are handled by the respective
+"Deliver" operation. Depending on the use case, "Deliver" spans can correlate
+with "Process" spans or other spans modelling processing operations.
+
+The operation modelled by the "Deliver" span does not strictly refer to
+receiving the message from intermediaries, but instead refers to the
+application receiving messages for processing. If messages are fetched from the
+intermediary and forwarded to the application in one go, the whole operation
+might be covered by a "Deliver" span. However, clients might pre-fetch messages
+from intermediaries and cache those messages, and only forward messages to the
+application at a later time. In this case, the operation of pre-fetching and
+caching should not be covered by the "Deliver" span.
+
+"Deliver" spans SHOULD be created for all messages obtained by or passed to the
+application for processing. "Deliver" spans MUST NOT be created for messages
```

Comment:
I don't think it's a feasible requirement to satisfy:

> "Deliver" spans MUST NOT be created for messages not forwarded to the application

E.g. Azure EventHubs SDK is multi-component and the Processor part of it calls into the Receiver package. Receiver package does not know if `receive` call is made by user or another SDK.

Same with Kafka: [poll call](https://www.tutorialkart.com/apache-kafka/kafka-consumer-with-example-java-application/) does not know if it's executed by the user or 3rd party [Spring Kafka integration](https://www.baeldung.com/spring-kafka#2-consuming-messages). 

So the message can be requested by another library, service mesh, or serverless runtime, but never forwarded to the application because of the bug, filter, etc 

I think the intent here can be described with something like: `Deliver` spans MUST be created for explicit requests to the messaging library and MUST NOT be created to trace pre-fetched or cached messages.

Reply:
Do you think this concern is rendered obsolete by having separate conventions/spans for push ("Deliver") and pull ("receive") scenarios?